TypeScript 2.4.1 causes compile errors.

There is some enhanced type checking introduced in TypeScript 2.4, generating errors in the current source code.  For example:
```
src/Recognizer.ts(73,4): error TS2322: Type 'Readonly<Map<string, number>>' is not assignable to type 'ReadonlyMap<string, number> | undefined'.
```

Also the stub decorators like `@NotNull` aren't declared properly to be used as parameter decorators.

--- a/benchmark/TestPerformance.ts
+++ b/benchmark/TestPerformance.ts
@@ -1665,7 +1665,7 @@ class DescriptiveErrorListener implements ParserErrorListener {
 	static INSTANCE: DescriptiveErrorListener =  new DescriptiveErrorListener();
 
 	@Override
-	syntaxError<T extends Token>(recognizer: Recognizer<T,any>, offendingSymbol: T, line: number, charPositionInLine: number, msg: string, e: RecognitionException): void {
+	syntaxError<T extends Token>(recognizer: Recognizer<T,any>, offendingSymbol: T | undefined, line: number, charPositionInLine: number, msg: string, e: RecognitionException): void {
 		if (!TestPerformance.REPORT_SYNTAX_ERRORS) {
 			return;
 		}
@@ -1685,7 +1685,7 @@ class DescriptiveLexerErrorListener implements ANTLRErrorListener<number> {
 	static INSTANCE: DescriptiveLexerErrorListener =  new DescriptiveLexerErrorListener();
 
 	@Override
-	syntaxError<T extends number>(recognizer: Recognizer<T,any>, offendingSymbol: T, line: number, charPositionInLine: number, msg: string, e: RecognitionException): void {
+	syntaxError<T extends number>(recognizer: Recognizer<T,any>, offendingSymbol: T | undefined, line: number, charPositionInLine: number, msg: string, e: RecognitionException): void {
 		if (!TestPerformance.REPORT_SYNTAX_ERRORS) {
 			return;
 		}
@@ -1702,11 +1702,11 @@ class DescriptiveLexerErrorListener implements ANTLRErrorListener<number> {
 }
 
 class SummarizingDiagnosticErrorListener extends DiagnosticErrorListener {
-	private _sllConflict: BitSet;
+	private _sllConflict: BitSet | undefined;
 	private _sllConfigs: ATNConfigSet;
 
 	@Override
-	reportAmbiguity(recognizer: Parser, dfa: DFA, startIndex: number, stopIndex: number, exact: boolean, ambigAlts: BitSet, configs: ATNConfigSet): void {
+	reportAmbiguity(recognizer: Parser, dfa: DFA, startIndex: number, stopIndex: number, exact: boolean, ambigAlts: BitSet | undefined, configs: ATNConfigSet): void {
 		if (TestPerformance.COMPUTE_TRANSITION_STATS && TestPerformance.DETAILED_DFA_STATE_STATS) {
 			let sllPredictions: BitSet =  this.getConflictingAlts(this._sllConflict, this._sllConfigs);
 			let sllPrediction: number =  sllPredictions.nextSetBit(0);
@@ -1729,7 +1729,7 @@ class SummarizingDiagnosticErrorListener extends DiagnosticErrorListener {
 	}
 
 	@Override
-	reportAttemptingFullContext(recognizer: Parser, dfa: DFA, startIndex: number, stopIndex: number, conflictingAlts: BitSet, conflictState: SimulatorState): void {
+	reportAttemptingFullContext(recognizer: Parser, dfa: DFA, startIndex: number, stopIndex: number, conflictingAlts: BitSet | undefined, conflictState: SimulatorState): void {
 		this._sllConflict = conflictingAlts;
 		this._sllConfigs = conflictState.s0.configs;
 		if (!TestPerformance.REPORT_FULL_CONTEXT) {

--- a/package.json
+++ b/package.json
@@ -58,7 +58,7 @@
     "source-map-support": "^0.4.3",
     "std-mocks": "^1.0.1",
     "typedoc": "^0.5.1",
-    "typescript": "^2.1.0"
+    "typescript": "^2.4.1"
   },
   "dependencies": {}
 }

--- a/src/Decorators.ts
+++ b/src/Decorators.ts
@@ -6,13 +6,13 @@
 export function NotNull(
 	target: any,
 	propertyKey: PropertyKey,
-	propertyDescriptor?: PropertyDescriptor) {
+	propertyDescriptor?: PropertyDescriptor | number) {
 }
 
 export function Nullable(
 	target: any,
 	propertyKey: PropertyKey,
-	propertyDescriptor?: PropertyDescriptor) {
+	propertyDescriptor?: PropertyDescriptor | number) {
 }
 
 export function Override(target: any,

--- a/src/DiagnosticErrorListener.ts
+++ b/src/DiagnosticErrorListener.ts
@@ -33,7 +33,10 @@ import { BitSet } from './misc/BitSet';
 import { DFA } from './dfa/DFA';
 import { Parser } from './Parser';
 import { ParserErrorListener } from './ParserErrorListener';
+import { RecognitionException } from './RecognitionException';
+import { Recognizer } from './Recognizer';
 import { SimulatorState } from './atn/SimulatorState';
+import { Token } from './Token';
 import { Override, NotNull } from "./Decorators";
 import { Interval } from "./misc/Interval";
 import { asIterable } from './misc/Stubs';
@@ -49,6 +52,18 @@ export class DiagnosticErrorListener implements ParserErrorListener {
 	 */
 	constructor(protected exactOnly: boolean = true) {
 		this.exactOnly = exactOnly;
+	}
+
+	@Override
+	syntaxError<T extends Token>(
+		/*@NotNull*/ recognizer: Recognizer<T, any>,
+		offendingSymbol: T | undefined,
+		line: number,
+		charPositionInLine: number,
+		/*@NotNull*/
+		msg: string,
+		e: RecognitionException | undefined): void
+	{
 	}
 
 	@Override

--- a/src/Recognizer.ts
+++ b/src/Recognizer.ts
@@ -70,7 +70,7 @@ export abstract class Recognizer<Symbol, ATNInterpreter extends ATNSimulator> {
 			}
 
 			intermediateResult.set("EOF", Token.EOF);
-			result = Object.freeze(intermediateResult);
+			result = intermediateResult;
 			Recognizer.tokenTypeMapCache.set(vocabulary, result);
 		}
 
@@ -91,7 +91,7 @@ export abstract class Recognizer<Symbol, ATNInterpreter extends ATNSimulator> {
 
 		let result: ReadonlyMap<string, number> | undefined = Recognizer.ruleIndexMapCache.get(ruleNames);
 		if (result == null) {
-			result = Object.freeze(Utils.toMap(ruleNames));
+			result = Utils.toMap(ruleNames);
 			Recognizer.ruleIndexMapCache.set(ruleNames, result);
 		}
 

--- a/src/tree/ParseTreeWalker.ts
+++ b/src/tree/ParseTreeWalker.ts
@@ -12,7 +12,7 @@ import { RuleNode } from "./RuleNode";
 import { ParserRuleContext } from "../ParserRuleContext";
 
 export class ParseTreeWalker {
-	walk(listener: ParseTreeListener, t: ParseTree): void {
+	walk<T extends ParseTreeListener>(listener: T, t: ParseTree): void {
 		let nodeStack: ParseTree[] = [];
 		let indexStack: number[] = [];
 

--- a/src/tree/xpath/XPathLexerErrorListener.ts
+++ b/src/tree/xpath/XPathLexerErrorListener.ts
@@ -12,7 +12,7 @@ import { RecognitionException } from "../../RecognitionException";
 
 export class XPathLexerErrorListener implements ANTLRErrorListener<number> {
 	@Override
-	syntaxError<T extends number>(recognizer: Recognizer<T, any>, offendingSymbol: T,
+	syntaxError<T extends number>(recognizer: Recognizer<T, any>, offendingSymbol: T | undefined,
 		line: number, charPositionInLine: number, msg: string,
 		e: RecognitionException): void {
 	}

--- a/tool/test/org/antlr/v4/test/runtime/typescript/TypeScript.test.stg
+++ b/tool/test/org/antlr/v4/test/runtime/typescript/TypeScript.test.stg
@@ -363,7 +363,7 @@ export class LeafListener implements TListener {
 >>
 
 WalkListener(s) ::= <<
-ParseTreeWalker.DEFAULT.walk(new LeafListener(), <s>);
+ParseTreeWalker.DEFAULT.walk\<TListener>(new LeafListener(), <s>);
 >>
 
 TreeNodeWithAltNumField(X) ::= <<

